design/operator-architecture.asciidoc

```diff
@@ -0,0 +1,89 @@
+
+* an opinionated way of installing and maintaining Keycloak installation on both Kubernetes and OpenShift.
+* an example of a recommended way of installing, configuring, managing custom extension and themes.
+* the thinnest possible layer on top of Keycloak Container and Keycloak Server.
+
+## Introduction
+
+Keycloak Operator was re-designed from the scratch by a small team withing RHMI and RHSSO groups. The implementation plan was described in the [Initial Operator Design Document](operator.md). At the end of a 3 sprint cycle, the team delivered a Level 4 Operator, that was pushed into the [Operatorhub](https://operatorhub.io/operator/keycloak-operator). Once the initial implementation was done, the Operator was handed over to the Keycloak/RHSSO Team.
+
+After maintaining the operator for several months, it became necessary to create a High Level Design (and a long-term vision) how all Keycloak components work together.
+
+## Layered architecture
+
+Keycloak ecosystem consists of 3 layers:
+
+* https://github.com/keycloak/keycloak[Keycloak Server]
+* https://github.com/keycloak/keycloak-containers/tree/master/server[Keycloak Container Image]
+* https://github.com/keycloak/keycloak-operator[Keycloak Operator]
+
+image::img/Keycloak-operator-layers.png[Keycloak ecosystem layers]
+
+Keycloak Operator uses Keycloak Container image, which in turn uses Keycloak Server bits. The goal is to implement features in the lowest possible layer, so that they get inherited by the all upper layers.
+
+.Feature implementation example
+****
+Let's consider a simple feature - adding JSON formatted logs based on an environmental variable.
+
+This feature can potentially be implemented in any of the 3 layers mentioned before.
+
+Implementing it in the Server doesn't make much sense as it will require manipulating `standalone.xml` file based on some environment property.
+
+It can also be implemented in the container as well as in the Operator. However, we push it down the stack, so the correct place for the implementation is the container.
+****
+
+## Components created by the Operator
+
+The diagram below presents the most important connections between components:
+
+image::img/Keycloak-operator-components.png[Keycloak Operator components]
+
+When Keycloak Operator spins up Keycloak installation, it creates:
+
+* `keycloak-db-secret` - used to store database username, password and other properties, such as external address (if used).
+* `credentials-<<CR Name>>` - Admin user credentials to log into Keycloak installation
+* `keycloak` StatefulSet with HA support
+* `keycloak-postgresql` - Responsible for spinning up Postgresql deployment
+* `keycloak-discovery` Service - used for `JDBC_PING` discovery
+* `keycloak` Service - Used for connecting Keycloak using HTTPS (HTTP is not allowed)
+* `keycloak-postgresql` - A service for connecting both internal and external (if used) database instance
+* `keycloak` Route or Ingress - Used for accessing Keycloak
```

Comment:
Might be worth listing that one is for an OpenShift install and the other for native Kubernetes

Reply:
Good point. Fixed.